File: 0871-keys-and-rooms/0871-keys-and-rooms.java
class Solution {
    int n;

    public void dfs(int node, List<List<Integer>> rooms, boolean[] vis){
        vis[node] = true;

        for(int next : rooms.get(node)){
            if(!vis[next]){
                dfs(next, rooms, vis);
            }
        }
    }

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        this.n = rooms.size();
        boolean[] vis = new boolean[n];
        vis[0] = true;

        for(int next : rooms.get(0)){
            dfs(next, rooms, vis);
        }

        for(int i=0; i<n; i++){
            if(!vis[i]){
                return false;
            }
        }

        return true;
    }
}

File: 0871-keys-and-rooms/0871-keys-and-rooms.py
class Solution(object):
    def canVisitAllRooms(self, rooms):
        """
        :type rooms: List[List[int]]
        :rtype: bool
        """
        visited = set()
        stack = []

        for i in rooms[0]:
            stack.append(i)
        visited.add(0)

        while stack:
            room = stack.pop()
            
            if room in visited:
                continue
            
            visited.add(room)
            
            for keys in rooms[room]:
                if keys not in visited:
                    stack.append(keys)

        return len(visited) == len(rooms)

        
        
